discovery/watcher.go

```diff
@@ -366,3 +367,84 @@ func (w *Watcher) getDataplaneFeatures() (map[string]bool, error) {
 
 	return features, nil
 }
+
+// watch blocks to wait for server set changes. This aborts on receiving an
+// error from the server, when the Watcher's context is cancelled, or (TODO)
+// when the Watcher is told to switch servers.
+//
+// This returns an addrSet containing the most recent set of servers. If the
+// addrSet is nil, then there were not changes seen to the server set. (This is
+// the case if the server watch stream is disabled.
+//
+// When this returns with an error, the current server should no longer be
+// considered "OK". This may return a non-nil addrSet and a non-nil error (it
+// will usually do this when the server watch stream is aborted).
+func (w *Watcher) watch() (*addrSet, error) {
+	current := w.currentServer.Load().(serverState)
+	if current.dataplaneFeatures["DATAPLANE_FEATURES_WATCH_SERVERS"] && !w.config.ServerWatchDisabled {
+		return w.watchStream()
+	} else {
+		return w.watchSleep()
+	}
+}
+
+// watchStream opens a gRPC stream to receive server set changes. This blocks
+// potentially forever.
+//
+// This may be aborted when the gRPC stream receives some error, when the
+// Watcher's context is cancelled, or (TODO) when the Watcher is told to switch
+// servers.
+//
+// If an error that aborts the gRPC stream, we pass the non-nil error back,
+// and the server is marked unhealthy elsewhere.
+func (w *Watcher) watchStream() (*addrSet, error) {
+	w.log.Debug("Watcher.watchStream")
+	client := pbserverdiscovery.NewServerDiscoveryServiceClient(w.conn)
+	serverStream, err := client.WatchServers(w.ctx, &pbserverdiscovery.WatchServersRequest{})
+	if err != nil {
+		w.log.Error("opening server watch stream", "err", err)
+		return nil, err
+	}
+
+	var set *addrSet
+	for {
+		// This blocks until there is a change from the server.
+		resp, err := serverStream.Recv()
+		if err != nil {
+			w.log.Error("unable to receive from server watch stream", "err", err)
+			return set, err
+		}
+
+		// The set of servers from the stream is the best known set of servers to use.
+		set := newAddrSet()
+		for _, srv := range resp.Servers {
+			addr, err := w.nodeToAddrFn(srv.Id, srv.Address)
+			if err != nil {
+				// failed to parse address. ignore this server.
+				w.log.Warn(err.Error())
+				continue
+			}
+			set.Put(OK, addr)
+		}
+	}
+}
+
+// watchSleep is used when the server watch stream is not supported.
+// It may be interrupted if we are told to switch servers.
+func (w *Watcher) watchSleep() (*addrSet, error) {
+	w.log.Debug("Watcher.watchSleep", "interval", w.config.ServerWatchDisabledInterval)
+
+	for {
+		select {
+		case <-w.ctx.Done():
+			return nil, w.ctx.Err()
+		case <-time.After(w.config.ServerWatchDisabledInterval):
+		}
+
+		// is the server still OK?
+		_, err := w.getDataplaneFeatures()
+		if err != nil {
+			return nil, err
+		}
```

Comment:
Thoughts on a periodic "ping" of the Consul servers when the server watch is disabled?

Without some kind of ping, we may not ever get feedback that the server is no longer ok: if callers are only using http clients and not making any gRPC requests.

I think my main concern is the amount of load this will cause, and what a reasonable default interval would be. And we probably want to add some jitter/randomness to the interval.

Reply:
Right, it's a forced switch on resource exhaustion